reference/opt/shaders-msl/comp/force-recompile-hooks.swizzle.comp

```diff
@@ -7,6 +7,7 @@ using namespace metal;
 
 struct spvAux
 {
+    uint vertexCount;
```

Comment:
Is this enough to deal with indirect draw calls?

Reply:
Now you've got me a bit concerned. In that case, we wouldn't necessarily be able to extract the `vertexCount` from the indirect buffer, especially if it's in device-private memory. We'd have to copy it to the auxiliary buffer, in which case we would no longer be able to use Metal's `setXxxBytes:` method; or, we'd have to pass the indirect buffer itself to the shader.

Reply:
Yes, I think the only real solution for indirect is to bind the indirect buffer as a `uint *`, not the indirect struct (because offset for vkCmdIndirectDraw can align to just 4 bytes). Not sure how to deal with difference between vkCmdDraw and vkCmdDrawIndexed, because they have different structs, but for indirect, it is not mandatory to support indexed draw calls I think ... The AuxBuffer could store the indirect draw offset and stride (for multi-draw-indirect) into the indirect buffer. This only be relevant for vertex output capture, so the extra overhead should be minimal.

I don't think we need any copies, that would be nightmare fuel.

Reply:
> The AuxBuffer could store the indirect draw offset and stride (for multi-draw-indirect) into the indirect buffer.

I'm not even sure we need that. Metal has methods for updating only the offset of a buffer resource:

```objective-c
[renderCmdEnc setVertexBufferOffset: 32 atIndex: 0];
[renderCmdEnc setFragmentBufferOffset: 0xf00 atIndex: 5];
[computeCmdEnc setBufferOffset: 48 atIndex: 31];
```

Since multi draws are unrolled anyway, we could just update the offset of the indirect buffer passed to the shader.

Reply:
The concern is alignment of those offsets. In Vulkan on nVidia for example, the alignment for an SSBO offset is 256 bytes or something like that. Does Metal have no such restrictions?

Reply:
SSBOs (buffers in the `device` address space) have natural alignment. Uniform buffers (in the `constant` address space) also have natural alignment on iOS. On Mac, they are aligned to 256 bytes.

Reply:
The relevant bit from the docs:

> For buffers in the *device* address space, align the offset to the data type consumed by the function (which is always less than or equal to 16 bytes).
>
> For buffers in the *constant* address space, align the offset to 256 bytes in macOS. In iOS, align the offset to the maximum of either the data type consumed by the function, or 4 bytes. A 16-byte alignment is safe in iOS if you don't need to consider the data type.

Reply:
I see. Then we shouldn't need any offset/stride for indirect command arguments.

Reply:
I've added something to deal with indirect draws. And yeah, this gets rid of the extra field from the aux buffer, rendering that entire discussion moot. :smile That discussion probably needed to happen anyway.